# Overview


In [45]:
import pandas as pd
import matplotlib.pyplot as plt

In [4]:
timeseries_data_path = '../dataset/raw/timeseries_data.csv'
well_data_path = '../dataset/raw/well_data.csv'

In [10]:
df_ts = pd.read_csv(timeseries_data_path,index_col=0,parse_dates=['date'])
df_ts.head()

well       date     value
0   well_9 1977-01-01  0.939394
1  well_28 1977-01-01  0.020087
2   well_4 1977-01-01  0.939394
3  well_12 1977-01-01  0.563636
4  well_17 1977-01-01  0.391414

In [11]:
df_ts.info()

<class 'pandas.core.frame.DataFrame'>
Index: 167891 entries, 0 to 167890
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   well    161491 non-null  object        
 1   date    167891 non-null  datetime64[ns]
 2   value   167891 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 5.1+ MB


In [58]:
df_ts[df_ts.well.isna()]

well       date      value
31      NaN 1977-01-01  75.879310
32      NaN 1977-01-01   5.753788
51      NaN 1977-01-01   0.276786
109     NaN 1977-01-01  35.953608
151     NaN 1977-01-01   0.779412
...     ...        ...        ...
167736  NaN 2018-01-01  11.538462
167777  NaN 2018-01-01   6.692641
167803  NaN 2018-01-01  20.800000
167833  NaN 2018-01-01   2.903614
167839  NaN 2018-01-01   2.319066

[6400 rows x 3 columns]

In [62]:
(pd.read_csv(timeseries_data_path,index_col=0,parse_dates=['date'])
 .dropna()
 )

well       date      value
0         well_9 1977-01-01   0.939394
1        well_28 1977-01-01   0.020087
2         well_4 1977-01-01   0.939394
3        well_12 1977-01-01   0.563636
4        well_17 1977-01-01   0.391414
...          ...        ...        ...
167886  well_259 2018-02-01   8.022222
167887  well_281 2018-02-01   4.453216
167888  well_270 2018-02-01  14.288775
167889  well_432 2018-02-01   9.398742
167890  well_364 2018-02-01   1.487923

[161491 rows x 3 columns]

In [14]:
df_ts.describe()

date          value
count                         167891  167891.000000
mean   1993-10-16 07:20:01.014944256      18.319189
min              1977-01-01 00:00:00       0.000000
25%              1984-10-01 00:00:00       1.071429
50%              1992-02-01 00:00:00       5.461725
75%              2001-12-01 00:00:00      17.437237
max              2018-02-01 00:00:00   12183.000000
std                              NaN      68.106097

In [36]:
df_grouped = df_ts.groupby('well')
summary_stats = df_grouped.describe()
summary_stats.sample(10)

date                                                      \
         count                           mean                  min   
well                                                                 
well_529   284  1994-08-16 11:04:13.521126784  1982-11-01 00:00:00   
well_354   127  1982-04-01 11:20:18.897637824  1977-01-01 00:00:00   
well_353   279  1988-08-20 09:32:54.193548416  1977-01-01 00:00:00   
well_259   206  2003-10-10 09:12:13.980582528  1990-01-01 00:00:00   
well_582   263  1995-05-02 04:44:42.889733888  1984-06-01 00:00:00   
well_510   431  2000-03-01 13:18:30.904872448  1982-04-01 00:00:00   
well_50    484  1997-09-15 15:34:12.892561920  1977-08-01 00:00:00   
well_3     275  1988-08-31 03:55:38.181818240  1977-04-01 00:00:00   
well_546   204  1991-05-15 14:14:07.058823552  1982-08-01 00:00:00   
well_402   157  1983-08-31 22:19:06.496815296  1977-03-01 00:00:00   

                                                                         \
                          25%                  50%                  75%   
well                                                                      
well_529  1988-09-23 12:00:00  1994-08-16 12:00:00  2000-07-08 18:00:00   
well_354  1979-08-16 12:00:00  1982-04-01 00:00:00  1984-11-16 00:00:00   
well_353  1982-10-16 12:00:00  1988-08-01 00:00:00  1994-05-16 12:00:00   
well_259  1994-04-08 12:00:00  1998-07-16 12:00:00  2013-10-24 06:00:00   
well_582  1989-11-16 00:00:00  1995-05-01 00:00:00  2000-10-16 12:00:00   
well_510  1991-03-16 12:00:00  2000-03-01 00:00:00  2009-02-15 00:00:00   
well_50   1987-08-24 06:00:00  1997-09-16 00:00:00  2007-10-08 18:00:00   
well_3    1982-12-16 12:00:00  1988-09-01 00:00:00  1994-05-16 12:00:00   
well_546  1986-10-24 06:00:00  1991-01-16 12:00:00  1996-04-08 12:00:00   
well_402  1980-06-01 00:00:00  1983-09-01 00:00:00  1986-12-01 00:00:00   

                                    value                                  \
                          max  std  count       mean       min        25%   
well                                                                        
well_529  2006-06-01 00:00:00  NaN  284.0   2.988442  0.000000   1.954031   
well_354  1987-07-01 00:00:00  NaN  127.0   5.000165  0.848485   1.660714   
well_353  2001-03-01 00:00:00  NaN  279.0  12.383599  0.000000   2.439566   
well_259  2018-02-01 00:00:00  NaN  206.0  40.956238  0.000000   3.406528   
well_582  2006-04-01 00:00:00  NaN  263.0  20.787318  0.000000   1.718021   
well_510  2018-02-01 00:00:00  NaN  431.0   7.721195  0.000000   4.377947   
well_50   2017-11-01 00:00:00  NaN  484.0  21.095983  0.000000   3.968229   
well_3    2000-02-01 00:00:00  NaN  275.0  24.604505  0.945000   7.256501   
well_546  2000-07-01 00:00:00  NaN  204.0  42.756795  0.000000  22.037106   
well_402  1990-03-01 00:00:00  NaN  157.0  15.919179  0.000000   7.175492   

                                                       
                50%        75%         max        std  
well                                                   
well_529   2.691809   3.782780   13.334630   1.619567  
well_354   2.533784   6.166676   32.938272   5.769314  
well_353   6.279221  11.861153  211.771429  22.089032  
well_259  12.565651  66.666667  273.026316  56.064003  
well_582   4.054054   9.269231  573.500000  66.710068  
well_510   6.471761   8.789210   97.600000   7.079822  
well_50    5.000000  35.628472  175.000000  26.802177  
well_3    19.338936  30.453447  209.250000  24.979002  
well_546  32.291176  50.655502  223.625000  42.181547  
well_402  10.660633  16.120000   99.000000  16.891452

In [34]:
summary_stats.describe()

value                                                              \
            count        mean         min         25%         50%         75%   
count  557.000000  557.000000  557.000000  557.000000  557.000000  557.000000   
mean   289.929982   20.921234    0.190098    7.918503   14.439853   24.286447   
std    131.643255   28.853106    1.391646   15.429708   24.505241   34.014846   
min     42.000000    0.056421    0.000000    0.000000    0.000000    0.070886   
25%    178.000000    4.699807    0.000000    0.661666    1.918979    4.183454   
50%    273.000000   11.350420    0.000000    2.937943    6.542781   11.581522   
75%    420.000000   25.019587    0.000000    6.864164   14.901840   32.483871   
max    494.000000  225.557787   17.631579  124.442398  252.000000  291.680576   

                                 
                max         std  
count    557.000000  557.000000  
mean     341.984685   32.354788  
std      856.233948   66.309816  
min        1.301370    0.122343  
25%       44.045833    5.310300  
50%      123.923077   16.021391  
75%      280.272727   32.191582  
max    12183.000000  782.846882

* We have $557$ wells.
* The wells do not have a common date range some starts 1977, others start at 1982.
* The timeseries is sample on a montly basis. 
* There are some discontinuity in the collection of measurments. the well time series can have missing values 
* Each well is assiciated with a time series with a length that ranges from $42$ to $494$ samples.
* The value of the time series starts from $0$ and can go as high as $121,83$.
* most of the values of ts are less than $300$.